### ISSUE TITLE
Issue with downloading large files

### ISSUE THREAD

**Issue by GitBoudewijn** [dsohler/oneclick-downloader#6]
Hey,

Thanks for the update. I have come across an issue that may be harder to solve.

When downloading multiple large (2-5MB) image files, some images end up not downloading correctly (Chrome on Windows). For example an image that's supposed to be 4MB is only 2MB after downloading and the bottom half is grey. According to Chrome the downloads are completed without errors.

I don't know if there's a way to check if a download is completed, or to bypass Chrome's download manager entirely?

One thing that could maybe help is if there was a setting to set the maximum number of concurrent downloads, since I don't think it occurs when downloading one file at a time.

What also would help is the ability to redownload and overwrite specific files from the last step. Right now when you click on an image in the last step it opens it in the extension screen. If you could add the **download** attribute to those links it would instead download them. I don't think this method allows you to overwrite files though.

Thanks!

**Comment by dsohler** [comment]
The issue is not basing on the file size but on the duration the download takes. I noticed it, too. When there are too many downloads that take too long long-running downloads are corrupt. I guess this happens because the downloads take too long and one of the invoked mechanisms (XHR, Chrome, the web server, etc.) run into a timeout-

Since XHR (which is [used](https://github.com/dsohler/oneclick-downloader/blob/master/oneclick-downloader/downloader/download.js#L20-L61) to get the image pages’ contents and directly put the content into the download function) is asynchronous by default and setting it to synchronous causes error log messages about the deprecation of doing so it is not an easy task to create a queue of downloads that will be processed one after another.

It would definitely be a long-term process to implement re-downloading of failed images. Checking downloads for validity isn’t directly possible as far as I know. Same with bypassing Chrome’s downloading mechanism. Extensions are very sandboxed and JavaScript never interacts directly with anything outside the sandbox.

**Comment by GitBoudewijn** [comment]
Hmm that's a big problem.

I don't fully understand how it works but couldn't you let the extension page handle the downloads? It appears that the extension page can access the links since it displays them in a list. So can't the extension page initiate the downloads as well?

**Comment by dsohler** [comment]
There are two main issues with that.

First: Yes, it would be possible to get all the links first and send them to Chrome for downloading them one after another, but that would result in a lot of not downloaded links because the direct URL is only valid for a defined period of time at almost any hosting service.

Second: Currently when selecting the links and click download it fires XHR requests to all of the links, gets the direct URL for the images and sends the result to Chrome for downloading.

Hooking into either of them is difficult with the current workflow.

**Comment by GitBoudewijn** [comment]
Yeah I meant getting all the unprocessed links first and then getting the direct links one by one when it's time.

What if when you click download it doesn't send requests to all of the links but just to the first link, waits for that to finish downloading, then sends a request to the second link etc.?

**Comment by dsohler** [comment]
That’s what I have in my mind, too. Currently the extension isn’t developed like that because it had a much smaller scope (20-30 images at each run). Now it’s too late to easily change it. 

Currently all links are get and in the next step all URLs are fetched asynchronously by XHR. The problem is that all of the invoked things can’t see each other in an easy way. I’ll try to figure it out when I have time again to not only do bugfixing but actively developing the extension again.
